packages/grafana-data/src/field/fieldState.ts

```diff
@@ -0,0 +1,148 @@
+import { DataFrame, Field, TIME_SERIES_FIELD_NAME, FieldType } from '../types';
+import { formatLabels } from '../utils/labels';
+
+/**
+ * Get an appropriate display title
+ */
+export function getFrameDisplayTitle(frame: DataFrame, index?: number) {
+  if (frame.name) {
+    return frame.name;
+  }
+
+  // Single field with tags
+  const valuesWithLabels = frame.fields.filter(f => f.labels !== undefined);
+  if (valuesWithLabels.length === 1) {
+    return formatLabels(valuesWithLabels[0].labels!);
+  }
+
+  // list all the
+  if (index === undefined) {
+    return frame.fields
+      .filter(f => f.type !== FieldType.time)
+      .map(f => getFieldTitle(f, frame))
+      .join(', ');
+  }
+
+  if (frame.refId) {
+    return `Series (${frame.refId})`;
+  }
+
+  return `Series (${index})`;
+}
+
+export function getFieldTitle(field: Field, frame?: DataFrame, allFrames?: DataFrame[]): string {
```

Comment:
There's something with the signature of this function that is hard to reason about. The field passed here should be part of one or several frames, so you should be able to figure out the field in question with rowIndex and fieldIndex, but I'm certainly missing something.